pennylane/measurements/__init__.py

```diff
@@ -11,8 +11,133 @@
+  :class:`MeasurementTransform` and a :func:`~pennylane.batch_transform` is that a batch transform is
+  tracked by the gradient transform, while a :class:`MeasurementTransform` process isn't.
+  See :class:`ClassicalShadowMP` for an example.
+
+.. note::
+
+    The :func:`SampleMeasurement.process_samples`, :func:`StateMeasurement.process_state` and
+    :func:`MeasurementTransform.process` functions should always have the same signature.
+
+.. note::
+
+    A measurement process can inherit from both :class:`SampleMeasurement` and
+    :class:`StateMeasurement` classes, defining the needed logic to process either samples or the
+    quantum state. See :class:`VarianceMP` for an example.
+
+Differentiation
+^^^^^^^^^^^^^^^
+In general, a :class:`MeasurementProcess` is differentiable with respect to a parameter if the domain of
+that parameter is continuous. When using the analytic method of differentiation the output of the
+measurement process must be a real scalar value for it to be differentiable.
+
+Creating custom measurements
+----------------------------
+A custom measurement process can be created by inheriting from any of the classes mentioned above.
+
+The following is an example for a sample-based measurement that computes the number of samples
+obtained of a given state:
+
+.. code-block:: python
+
+    import pennylane as qml
+    from pennylane.measurements import SampleMeasurement
+
+    class CountState(SampleMeasurement):
+        def __init__(self, state: str):
+            self.state = state  # string identifying the state e.g. "0101"
+            wires = list(range(len(state)))
+            super().__init__(wires=wires)
+
+        def process_samples(self, samples, wire_order, shot_range, bin_size):
+            counts_mp = qml.counts(wires=self._wires)
+            counts = counts_mp.process_samples(samples, wire_order, shot_range, bin_size)
+            return counts.get(self.state, 0)
+
+        def __copy__(self):
+            return CountState(state=self.state)
+
+.. note::
+
+    The ``__copy__`` method needs to be overriden when new arguments are added into the ``__init__``
+    method.
+
+The measurement process in this example uses the :func:`~pennylane.counts` function, which is a
+measurement process which returns a dictionary containing the number of times each quantum
+state has been sampled.
+
+We can now execute the new measurement in a :class:`~pennylane.QNode`. Let's use a simple circuit
+so that we can verify our results mathematically.
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=1, shots=10000)
+
+    @qml.qnode(dev)
+    def circuit(x):
+        qml.RX(x, wires=0)
+        return CountState(state="1")
+
+.. note::
+
+    In PennyLane we use functions to define measurements (e.g. :func:`counts`). These
+    functions will return an instance of the corresponding measurement process
+    (e.g. :class:`CountsMP`).
```

Comment:
Is there any other reason to do this? One could wonder what is the point then 🤔 I would put this further above to clarify it from the beginning and also it may be confusing since here you are actually not doing that but directly instantiating the class.

Reply:
Just for design purposes. Added comment